src/Compilers/CSharp/Portable/Symbols/Source/ExtensionGroupingInfo.cs

```diff
@@ -0,0 +1,504 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Runtime.InteropServices;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.Emit;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class ExtensionGroupingInfo
+    {
+        private readonly Dictionary<string, MultiDictionary<string, SourceNamedTypeSymbol>> _groupingMap;
```

Comment:
nit: Consider documenting. If I understood correctly, the outer string key is a grouping name, the inner string key is a marker name, and the named types are the extension declarations
nit: it may be good to extract the computation logic that creates the dictionary and `ExtensionGroupingInfo` from `GetExtensionGroupingInfo` and move it into `ExtensionGroupingInfo` so the logic constructing this dictionary is closer to the dictionary's definition.